packages/mdl-textfield/README.md

```diff
@@ -0,0 +1,285 @@
+indicate to assistive devices that the display of the help text is dependent on the interaction with
+the input element.
+
+When using our vanilla JS component, if it sees that the input element has an `aria-controls`
+attribute, it will look for an element with the id specified and treat it as the text field's help
+text element, taking care of adding/removing `aria-hidden` and other a11y attributes. This can also
+be done programmatically, which is described below.
+
+### Validation
+
+MDL Textfield provides validity styling by using the `:invalid` and `:required` attributes provided
+by HTML5's form validation API.
+
+```html
+<div class="mdl-textfield">
+  <input type="password" id="pw" class="mdl-textfield__input" required minlength=8>
+  <label for="pw" class="mdl-textfield__label">Password</label>
+</div>
+```
+
+An input's validity is checked via `checkValidity()` on blur, and the styles are updated
+accordingly. When using the `required` attribute, an asterisk will be automatically appended to the
+label text, as per the spec.
+
+Help text can be used to provide additional validation messages. Use
+`mdl-textfield-helptext--validation-msg` to provide styles for using the help text as a validation
+message. This can be easily combined with `mdl-textfield-helptext--persistent` to provide a robust
+UX for client-side form field validation.
+
+```html
+<div class="mdl-textfield">
+  <input required minlength=8 type="password" class="mdl-textfield__input" id="pw"
+         aria-controls="pw-validation-msg">
+  <label for="pw" class="mdl-textfield__label">Choose password</label>
+</div>
+<p class="mdl-textfield-helptext
+          mdl-textfield-helptext--persistent
+          mdl-textfield-helptext--validation-msg"
+   id="pw-validation-msg">
+  Must be at least 8 characters long
+</p>
+```
+
+### Using the JS component
+
+MDL Textfield ships with Component / Foundation classes which are used to provide a full-fidelity
+Material Design text field component.
+
+#### Including in code
+
+##### ES2015
+
+```javascript
+import MDLTextfield, {MDLTextfieldFoundation} from 'mdl-textfield';
+```
+
+##### CommonJS
+
+```javascript
+const mdlTextfield = require('mdl-textfield');
+const MDLTextfield = mdlTextfield.default;
+const MDLTextfieldFoundation = mdlTextfield.MDLTextfieldFoundation;
+```
+
+##### AMD
+
+```javascript
+require(['path/to/mdl-textfield'], mdlTextfield => {
+  const MDLTextfield = mdlTextfield.default;
+  const MDLTextfieldFoundation = mdlTextfield.MDLTextfieldFoundation;
+});
+```
+
+##### Global
+
+```javascript
+const MDLTextfield = mdl.Textfield;
+const MDLTextfieldFoundation = mdl.Textfield.MDLTextfieldFoundation;
+```
+
+#### Fully-automatic: DOM Rendering + Initialization
+
+```javascript
+const label = document.createElement('label');
+label.textContent = 'My label text';
+const root = MDLTextfield.buildDom({
+  id: 'my-textfield',
+  helptextId: 'my-textfield-helptext',
+  labelElement: label
```

Comment:
We went this approach because we were afraid of things like: "oh, I want to use `<em>` elements in my label, so can you use innerHTML instead of innerText?" and then we have to deal with escaping issues etc. etc.

We're actually in the process of getting rid of `buildDom()` (#4806) so I'm gonna just nuke this